lib/blocks.php

```diff
@@ -295,36 +295,48 @@ function gutenberg_experimental_apply_classnames_and_styles( $block_content, $bl
  * Add CSS classes and inline styles for colors to the incoming attributes array.
  * This will be applied to the block markup in the front-end.
  *
- * @param  array $attributes comprehensive list of attributes to be applied.
- * @param  array $block_attributes block attributes.
- * @param  array $supports style features the block attributes.
+ * @param  array  $attributes comprehensive list of attributes to be applied.
+ * @param  array  $block_attributes block attributes.
+ * @param  array  $supports style features the block attributes.
+ * @param  object $block_type registered block type class.
  * @return array Colors CSS classes and inline styles.
  */
-function gutenberg_experimental_build_css_colors( $attributes, $block_attributes, $supports ) {
+function gutenberg_experimental_build_css_colors( $attributes, $block_attributes, $supports, $block_type ) {
 	// Text Colors.
 	// Check support for text colors.
 	if ( in_array( 'color', $supports, true ) ) {
-		$has_named_text_color  = array_key_exists( 'textColor', $block_attributes );
-		$has_custom_text_color = isset( $block_attributes['style']['color']['text'] );
+		$has_named_text_color          = array_key_exists( 'textColor', $block_attributes );
+		$has_custom_text_color         = isset( $block_attributes['style']['color']['text'] );
+		$has_named_text_color_default  = isset( $block_type->attributes['textColor']['default'] );
```

Comment:
Could you elaborate more on what the difference is between `$block_attributes` and `$block_type->attributes`?

Reply:
`block_attributes` are the attributes of the saved instance of the block.  `$block_type` is how that type of block is originally registered, and `$block_type->attributes` will be where initially defined attributes and default values live.

Reply:
Ahhhh. I see. `$block_attributes` are the stateful attributes that may or may not have been modified by users. 

It sounds like`$block_type` and it's attributes are used to create the initial instance of a block. Let me know if that's not quite right, but otherwise, thanks for the clarification! 🙏

Reply:
> It sounds like`$block_type` and it's attributes are used to create the initial instance of a block. Let me know if that's not quite right, but otherwise, thanks for the clarification! 🙏

This is mostly correct, but `$block_type` doesn't have attributes so much as it specifies a map of attributes for its future block instances. Think of `$block_type->attributes` more as a schema. For example:

```json
// blockType.attributes
{
	"value": {
		"type": "string",
		"source": "html",
		"selector": "blockquote",
		"multiline": "p",
		"default": ""
	},
	"align": {
		"type": "string"
	}
}
```

```json
// block.attributes
{
	"value": "<p>Hello</p>",
}
```

Reply:
Thanks for elaborating @mcsf!